### PR TITLE
Add lookup_methods endpoint

### PR DESCRIPTION
**Pull request by lok52**
Add `lookup-methods` endpoint for solidity contracts. Internally PoC logic from `sol-abi-jumps` library is used.

### CODE REVIEW

File: smart-contract-verifier/smart-contract-verifier-proto/proto/v2/api_config_http.yaml [hunk 1]
```diff
@@ -16,6 +16,11 @@ http:
     - selector: blockscout.smartContractVerifier.v2.SolidityVerifier.ListCompilerVersions
       get: /api/v2/verifier/solidity/versions
 
+    - selector: blockscout.smartContractVerifier.v2.SolidityVerifier.LookupMethods
+      post: /api/v2/verifier/solidity/lookup-methods
```
[the remark this answers is not in the document]
Comment (lok52) [reply smart-contract-verifier/smart-contract-verifier-proto/proto/v2/api_config_http.yaml]: Fixed

File: smart-contract-verifier/smart-contract-verifier-proto/proto/v2/smart-contract-verifier.proto [hunk 2]
```diff
@@ -218,3 +220,19 @@ message ListCompilerVersionsResponse {
   /// Compiler versions available
   repeated string compiler_versions = 1;
 }
+
+message LookupMethodsRequest {
+  string bytecode = 1;
+  string abi = 2;
+  string source_map = 3;
+  map<uint32, string> file_id_map = 4;
```
[the remark this answers is not in the document]
Comment (lok52) [reply smart-contract-verifier/smart-contract-verifier-proto/proto/v2/smart-contract-verifier.proto]: Fixed

File: smart-contract-verifier/smart-contract-verifier/src/lookup_methods/disassemble.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+impl Display for DisassembledOpcode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.args.is_empty() {
+            write!(f, "{}", self.operation.name)
+        } else {
+            write!(f, "{} {:?}", self.operation.name, self.args)
+        }
+    }
+}
```
Comment (rimrakhimov) [inline smart-contract-verifier/smart-contract-verifier/src/lookup_methods/disassemble.rs:29]: Nit: just interested: why have the same Debug and Display implementations? Can we just derive Debug implementation? Or at least to implement one through another if they are the same anyway

**Review by rimrakhimov — APPROVED** [review]

File: smart-contract-verifier/smart-contract-verifier/src/lookup_methods/disassemble.rs (hunk 3 above)
Comment (lok52) [reply smart-contract-verifier/smart-contract-verifier/src/lookup_methods/disassemble.rs]: Forgot to clean up this after testing. Removed the Display impl

File: smart-contract-verifier/smart-contract-verifier-server/src/types/lookup_methods.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+#[derive(Wrapper, From, Clone, Debug, PartialEq)]
+pub struct LookupMethodsResponseWrapper(proto::LookupMethodsResponse);
+
+impl From<LookupMethodsResponse> for LookupMethodsResponseWrapper {
+    fn from(response: LookupMethodsResponse) -> Self {
+        Self(proto::LookupMethodsResponse {
+            methods: response
+                .methods
+                .iter()
```
Comment (rimrakhimov) [inline smart-contract-verifier/smart-contract-verifier-server/src/types/lookup_methods.rs:42]: Nit: allows to get rid of clones in the map
```suggestion
                .into_iter()
```

File: smart-contract-verifier/smart-contract-verifier-proto/proto/v2/smart-contract-verifier.proto [hunk 5]
```diff
@@ -218,3 +220,19 @@ message ListCompilerVersionsResponse {
...
+  string abi = 2;
+  string source_map = 3;
+  map<uint32, string> file_ids = 4;
+}
+
+message LookupMethodsResponse {
+  message Method {
+    string file_name = 1;
+    int32 file_offset = 2;
+    int32 length = 3;
```
Comment (rimrakhimov) [inline smart-contract-verifier/smart-contract-verifier-proto/proto/v2/smart-contract-verifier.proto:235]: Nit: maybe change to `uint32`? In general they are completely compatible types; `uint` may just convey better sematical meaning

File: smart-contract-verifier/smart-contract-verifier/src/lookup_methods/find_methods.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+pub fn find_methods(request: LookupMethodsRequest) -> LookupMethodsResponse {
+    let methods = parse_selectors(request.abi);
+    let opcodes = disassemble_bytecode(&request.bytecode);
+    let opcodes = opcodes.as_slice();
+
+    let methods = methods
+        .into_iter()
+        .filter_map(|(func_sig, selector)| {
+            let func_index = match find_src_map_index(&selector, opcodes) {
```
Comment (rimrakhimov) [inline smart-contract-verifier/smart-contract-verifier/src/lookup_methods/find_methods.rs:29]: Nit: then the `opcodes` could be a vector (and there would be no need in `opcodes.as_slice()` line)
```suggestion
            let func_index = match find_src_map_index(&selector, &opcodes) {
```

File: smart-contract-verifier/smart-contract-verifier/src/lookup_methods/find_methods.rs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            ) {
+                Ok(m) => m,
+                Err(err) => {
+                    tracing::warn!(func_sig, err = err.to_string(), "failed to parse method");
+                    return None;
+                }
+            };
+            Some((hex::encode(selector), method))
+        })
+        .collect::<BTreeMap<String, Method>>();
```
Comment (rimrakhimov) [inline smart-contract-verifier/smart-contract-verifier/src/lookup_methods/find_methods.rs:52]: Nit: Just to make sure, that such explicit types declaration is intentional and you know that just a `collect()` will work here as well
```suggestion
        }).collect();
```

**Review by rimrakhimov — APPROVED** [review]
Looks great

**Review by rimrakhimov — APPROVED** [review]

File: smart-contract-verifier/smart-contract-verifier-server/tests/lookup_methods.rs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    let (request, methods) = process_compiler_output(&output, &test_case.contract_name).unwrap();
+    let res: LookupMethodsResponse = send_post_request(&base, ROUTE, &request).await;
+
+    // Make sure we extracted all methods
+    assert_eq!(
+        methods.values().collect::<BTreeSet<&String>>(),
+        res.methods.keys().collect::<BTreeSet<&String>>()
+    );
```
Comment (sevenzing) [inline smart-contract-verifier/smart-contract-verifier-server/tests/lookup_methods.rs:121]: you check that all methods are present, but you skip content of method location, what if we calculate jumps wrong?

I suggest to add expected methods location to config.json of every test

Comment (lok52) [reply smart-contract-verifier/smart-contract-verifier-server/tests/lookup_methods.rs:121]: Will consider adding it in the latter PR

**Pull request merged**